### DISCUSSION TITLE
API rules and filters - multi relation field with joins

### DISCUSSION THREAD

**Discussion by Furminto** [pocketbase/pocketbase#6119] [Q&A] [answered]
Hello!

First of all I love the project, and big thanks for this, it is amazing.

![uml_diagram](https://github.com/user-attachments/assets/e196bce0-fcbb-48c5-b13e-2350e8b03543)

I want to implement a company, team system for my application and I have these collections:
- user
- company
- userCompanies: connect the users and the companies, and has a role (admin or user)
- teams: one company, one leader(user), and multiple members(user)

pb version: 0.23.5

Beyond the enabling filtering rules (e.g. api caller is admin, or the owner) I want to check if the team leader belongs to the company (there is an userCompanies entity where userId and companyId are linked), and check the same for the members.

The owner can be written down very nicely:
`@request.body.leader.userCompanies_via_user.company.id ?= company.id`

But for the members even with the each modifier I cannot write the right rule. I tried a few things, maybe the best this is, but this is not working properly:
`@request.body.members:each.userCompanies_via_user.company.id ?= company.id`

I can write a record validation for this, which works, but I wonder is it possible to write a rule for this?

```
onRecordValidate((e) => {
    const companyId=  e.record.getString("company")
    const queryResult = arrayOf(new DynamicModel({
        user: "",
    }))
    e.app.db().newQuery("SELECT user FROM userCompanies WHERE company = {:companyId}")
    .bind({
        "companyId": companyId,
    }).all(queryResult)
    const companyMembers = queryResult.map((r) => r.user)
    const members = e.record.getStringSlice("members")
    const isEveryoneCompanyMember = members.every((member) => { 
        return companyMembers.includes(member) 
    })
    if(!isEveryoneCompanyMember){
        throw new ValidationError("invalid_memberrs", "there is invalid member")
    }
    e.next()
}, "teams")
```

Do I have a wrong model, or what should I do to make it more cleaner?

thanks, have a nice day :)

**Comment by ganigeorgiev** [comment, marked as answer]
`:each` (and the other modifiers) cannot be chained. Following directly with dot-notation should work, aka:

```js
@request.body.members.userCompanies_via_user.company ?= company
```

With that said, I'm not sure that I understand what is the intended behavior. Note that the above rule doesn't make any checks about the current authenticated user (that would be the `@request.auth.*`), only against the user(s) submitted in the body (assuming that by _"Beyond the enabling filtering rules"_ means that you have extra checks for this?). And additionally it also allows:
- members with different companies
- one member to add or remove other members (_if the member is allowed to modify the team_)

You can use the "match-all" equal operator (aka. just `=`) to resolve the first point but note that in this case this will require for every member and for each of their userCompanies relations to be related to a company satisfying your condition, which could fail if a single member is allowed to have more than one company. There is no partial "match-all"+"any/at-least-one" type of support for the same operand at the moment, if that's what you want.

If that's not the intended behavior and you don't want to use the hooks, then you'll have to reorganize your collections and use a more traditional SQL structure, for example: remove the multiple members (and maybe the leader) `relation` field and instead create a junction collection with 2 single relation fields pointing back to the teams and users (similar to userCompanies).

**Reply by Furminto** [reply]
Thanks for the quick reply, the model change which is the answer for me I think (junction collection).

There are irrelevant parts of the rule which I have omitted for simplicity in this case, e.g. the owner has access `@request.auth.id = company.owner.id`.
